processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -109,6 +271,83 @@
     return new GenericIndexed<T>(theBuffer.asReadOnlyBuffer(), strategy, allowReverseLookup);
   }
 
+  public static <T> GenericIndexed<T> read(ByteBuffer buffer, ObjectStrategy<T> strategy)
+  {
+    byte versionFromBuffer = buffer.get();
+
+    if (VERSION_ONE == versionFromBuffer) {
+      return createVersionOneGenericIndexed(buffer, strategy);
+    } else if (VERSION_TWO == versionFromBuffer) {
+      throw new IAE(
+          "use read(ByteBuffer buffer, ObjectStrategy<T> strategy, SmooshedFileMapper fileMapper) to read version 2 indexed.",
+          versionFromBuffer
+      );
+    }
+    throw new IAE("Unknown version[%s]", versionFromBuffer);
+  }
+
+  private static <T> GenericIndexed<T> createVersionOneGenericIndexed(ByteBuffer buffer, ObjectStrategy<T> strategy)
+  {
+    boolean allowReverseLookup = buffer.get() == 0x1;
+    int size = buffer.getInt();
+    ByteBuffer bufferToUse = buffer.asReadOnlyBuffer();
+    bufferToUse.limit(bufferToUse.position() + size);
+    buffer.position(bufferToUse.limit());
+
+    return new GenericIndexed<T>(
+        bufferToUse,
+        strategy,
+        allowReverseLookup
+    );
+  }
+
+  public static <T> GenericIndexed<T> read(ByteBuffer buffer, ObjectStrategy<T> strategy, SmooshedFileMapper fileMapper)
+  {
+    byte versionFromBuffer = buffer.get();
+
+    if (VERSION_ONE == versionFromBuffer) {
+      return createVersionOneGenericIndexed(buffer, strategy);
+    } else {
+      if (VERSION_TWO == versionFromBuffer) {
+
+        if (fileMapper == null) {
+          throw new IAE("SmooshedFileMapper can not be null for version 2.");
+        }
+        boolean allowReverseLookup = buffer.get() == 0x1;
+        int logBaseTwoOfElementsPerValueFile = buffer.getInt();
+        int numWritten = buffer.getInt();
+        String columnName;
+
+        List<ByteBuffer> valueBuffersToUse;
+        ByteBuffer headerBuffer;
+        try {
+          columnName = new SerializerUtils().readString(buffer);
+          valueBuffersToUse = Lists.newArrayList();
+
+          for (int i = 0; i <= numWritten / (1 << logBaseTwoOfElementsPerValueFile); i++) {
```

Comment:
numWritten can be greater than or equals to `1 << logBaseTwoOfElementsPerValueFile`, my assumption here is all the value files will be of same size except the last one. For an example say numWritten is 15 and  logBaseTwoOfElementsPerValueFile=3, means total 8 elements in first value file and second only has 7 elements. `numWritten / (1 << logBaseTwoOfElementsPerValueFile)` gives me 1 and loop starts from 0, so it will create 2 value files.
Yes, `numWritten / (1 << logBaseTwoOfElementsPerValueFile)` is the same as `numWritten >> logBaseTwoOfElementsPerValueFile`. Will change it to use `>>` for devision.